galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,571 @@
+# relative to -Y_pl, and of f1 relative to +X_pl.
+theta_fpa = 32.5*galsim.degrees
+
+# File with SIP coefficients.
+sip_filename = os.path.join(galsim.meta_data.share_dir, 'sip_422.txt')
+
+def getWCS(PA, ra=None, dec=None, pos=None, PA_is_FPA=False, as_header=False):
+    """
+    This routine gets a list of WCS, one for each of the WFIRST SCAs (Sensor Chip Array, the
+    equivalent of a chip in an optical CCD).  Since the WFIRST SCAs are labeled 1-18, the zeroth
+    list item is simply None.  Use of this routine requires that GalSim be able to access some
+    software that can handle TAN-SIP style WCS (either Astropy, starlink.Ast, or WCSTools).
+
+    The user must specify a position for the center of the focal plane array (either as (ra, dec),
+    or a CelestialCoord `pos`) and the orientation.
+
+    For this routine, we define several coordinate systems for WFIRST.  The diagram
+    located on the GalSim wiki,
+    https://github.com/GalSim-developers/GalSim/wiki/GalSim-WFIRST-module-diagrams
+    will be useful in understanding the description of these coordinate systems below.
+
+    Observatory coordinate system: +X_obs points along the boresight into the sky, +Z_obs points
+    towards the Sun in the absence of a roll offset, +Y_obs makes a right-handed system.
+
+    Payload coordinate system: +X_pl points along -Y_obs, +Y_pl points along +Z_obs, +Z_pl points
+    along -X_obs (back towards observer).
+
+    Wide field imager (WFI) focal plane assembly (FPA) coordinate system: This is defined by a
+    left-handed system f1, f2, that is rotated by an angle `theta_fpa` with respect to the payload
+    axes.  +f1 points along the long axis of the focal plane, transverse to the radius from the
+    telescope optic axis.  +f2 points radially out from the telescope optic axis, along the narrow
+    dimension of the focal plane.  If +f2 points North, then +f1 points East.  `theta_fpa` is a
+    positive CCW rotation of the f2 axis relative to -Y_pl, and of f1 relative to +X_pl.  In terms
+    of focal plane geometry, if +Y_fp is pointing North, then SCAs 3 and 12 will be at highest
+    declination, 8 and 17 at the lowest.  +Y_fp is aligned with the short axis of the focal plane
+    array.
+
+    There is also a detector coordinate system (P1, P2).  +P1 and +P2 point along the fast- and
+    slow-scan directions of the pixel readout, respectively.
+
+    So, for reference, if the boresight is pointed at RA=90, DEC=0 on March 21st (Sun at vernal
+    equinox), then +X_obs points at (RA,DEC)=(90,0), +Y_obs points North, and +Z_obs points at the
+    Sun.  The payload coordinates are +X_pl points South, -Y_pl points East.  Finally, the FPA
+    coordinate system is defined by +f2 being at a position angle 90+theta_fpa east of North.  If
+    the observatory +Y axis is at a position angle `pa_obsy` East of North, then the focal plane
+    (+f2) is at a position angle pa_fpa = pa_obsy + 90 + theta_fpa.
```

Comment:
OK.  I dug into this a little more, and here's the deal:

- That sentence in the docstring implies that things are a bit more deterministic than they actually are.  In other words, that's the optimal case (which results in the solar panels pointing directly at the sun), but some amount of rotation with respect to that orientation is allowed.

- One could, in principle, take an input date and fairly easily calculate the optimal PA.  I mean, I haven't done it but from the documentation they gave me (which is now in the docstring) it would be easy.

- I asked if they have code that calculates the allowed range of PA about the optimal one, and apparently it's very complicated and not in a good form for us to use.

- I could imagine two scenarios in practice.  The first is that the survey or the observer says "I want the following RA, Dec, PA" - and let the scheduler work out what dates that combination of RA, Dec, PA is allowed.  (In some cases they won't be allowed at all, in other cases there will be a wide or narrow range or dates in which it is allowed.)  In this scenario we would want people to be able to specify RA, Dec, PA.  

- The other scenario is that the person says "I want to go stare at this RA, Dec on this date, and don't care much about the PA so just do whatever works best."  In that case, we want the code to know how to calculate the optimal PA for a given date.

I honestly can't decide what would be the best thing to do here.  One option for the near term is to let the observer choose RA, Dec, PA, and simply state in the docstring that PA is actually constrained to some narrow range based on the date.  Then provide a helper routine they can use to calculate the optimal PA for a given date and ```world_pos```, or ignore, depending on how much they care about these details.  Later once there is reliable code to calculate the range of allowed PA we could incorporate that, but I have no idea when that will be. What do you think, given this additional info?